tests/integration/test_reward_integration.py

```diff
@@ -0,0 +1,139 @@
+    chain.mine(timedelta=60)
+
+    # Check reward per token calculation is accurate
+    reward_per_token_stored = multi.rewardData(reward_token)["rewardPerTokenStored"]
+    reward_rate = reward_amount // 60
+    time_max = multi.lastTimeRewardApplicable(reward_token)
+    time_min = multi.rewardData(reward_token)["lastUpdateTime"]
+    interval = time_max - time_min
+    rpt_calc = (interval * 10 ** 18 * reward_rate) // (amount1 + amount2)
+    rpt = multi.rewardPerToken(reward_token)
+    assert reward_per_token_stored + rpt_calc == rpt
+
+    # Check earning calculation is accurate
+    calc_earnings = (amount1 * rpt) // (10 ** 18)
+    act_earnings = multi.earned(bob, reward_token)
+    assert calc_earnings == act_earnings
+
+    # Account for the amount already stored
+    prepaid = multi.userRewardPerTokenPaid(charlie, reward_token)
+
+    calc_earnings2 = (amount2 * (rpt - prepaid)) // (10 ** 18)
+    act_earnings2 = multi.earned(charlie, reward_token)
+    assert calc_earnings2 == act_earnings2
+
+
+# Reward per token accurate?
+@given(amount=strategy("uint256", max_value=(10 ** 18), exclude=0))
+def test_reward_per_token(multi, alice, bob, reward_token, amount, chain, base_token):
+    reward_token.approve(multi, 10 ** 19, {"from": alice})
+    multi.setRewardsDistributor(reward_token, alice, {"from": alice})
+    multi.notifyRewardAmount(reward_token, 10 ** 10, {"from": alice})
+
+    init_rpt = multi.rewardPerToken(reward_token)
+    assert init_rpt == 0
+
+    base_token.approve(multi, amount, {"from": bob})
+    multi.stake(amount, {"from": bob})
+
+    chain.mine(timedelta=100)
+    final_rpt = multi.rewardPerToken(reward_token)
+
+    assert final_rpt // (10 ** 18) == multi.earned(bob, reward_token) // amount
+
+
+# Rewards struct updates as expected
+@given(amount=strategy("uint256", min_value=(10 ** 10), max_value=(10 ** 16), exclude=0))
+def test_rewards_update(multi, alice, reward_token, amount, chain, base_token):
+    reward_token.approve(multi, amount, {"from": alice})
+    multi.setRewardsDistributor(reward_token, alice, {"from": alice})
+    multi.notifyRewardAmount(reward_token, amount, {"from": alice})
+    multi.stake(amount, {"from": alice})
+
+    rewards = []
+    rewards.append(multi.rewardData(reward_token))
+    chain.mine(timedelta=60)
+    for i in range(1, 5):
+        reward_token.approve(multi, amount, {"from": alice})
+        multi.notifyRewardAmount(reward_token, amount, {"from": alice})
+        multi.stake(amount, {"from": alice})
+        chain.mine(timedelta=60)
+
+        rewards.append(multi.rewardData(reward_token))
+        curr = rewards[i]
+        last = rewards[i - 1]
+
+        assert last["periodFinish"] < curr["periodFinish"]
+        assert last["lastUpdateTime"] < curr["lastUpdateTime"]
+        assert last["rewardPerTokenStored"] < curr["rewardPerTokenStored"]
+
+
+@given(amount=strategy("uint256", min_value=0, max_value=1e77, exclude=0))
```

Comment:
Hypothesis generates values using a technique called [shrinking](https://hypothesis.works/articles/integrated-shrinking/), which means that it tends to start with large values and work towards smaller ones. The majority of the tests will happen in the lower end of the range. In general you want to avoid using overly large data sets, otherwise the test becomes ineffective.  In this case, I would wager 99% of the values used here will be less than `1e57`.

A test like this is better suited to simple parametrization, with one test where you target the exact upper and lower bounds of the specific behaviour, and a 2nd where you target just beyond those bounds.